packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -145,66 +168,407 @@ String getGradleVersionForAndroidPlugin(Directory directory, Logger logger) {
+  if (isWithinVersionRange(javaV, min: maxSupportedJavaVersion, max: '100.100', inclusiveMin: false)) {
+    // Assume versions Java versions newer than [maxSupportedJavaVersion]
+    // required a higher gradle version.
+    final bool validGradle =
+        isWithinVersionRange(gradleV, min: _maxKnownAndSupportedGradleVersion, max: '100.00');
+    logger.printWarning('Newer than known valid Java version ($javaV), gradle ($gradleV).'
+        '\n Treating as valid configuration.');
+    return validGradle;
+  }
+
+  // Begin known Java <-> Gradle evaluation.
+  if (isWithinVersionRange(javaV, min: '19', max: '20', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '7.6', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '18', max: '19', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '7.5', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '17', max: '18', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '7.3', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '16', max: '17', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '7.0', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '15', max: '16', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '6.7', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '14', max: '15', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '6.3', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '13', max: '14', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '6.0', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '12', max: '13', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '5.4', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '11', max: '12', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '5.0', max: _maxKnownAndSupportedGradleVersion);
+  }
+  // 1.11 is a made up java version to cover everything in 1.10.*
+  if (isWithinVersionRange(javaV, min: '1.10', max: '1.11', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '4.7', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '1.9', max: '1.10', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '4.3', max: _maxKnownAndSupportedGradleVersion);
+  }
+  if (isWithinVersionRange(javaV, min: '1.8', max: '1.9', inclusiveMax: false)) {
+    return isWithinVersionRange(gradleV,
+        min: '2.0', max: _maxKnownAndSupportedGradleVersion);
+  }
```

Comment:
Optional: this is very verbose and repetitive; it's hard to see the logic among the versions (e.g., if one of the calls switched `javaV` or `gradleV` we would probably never notice). If we can use `pub_semver` (see other comment) this could all be a `Map<VersionRange, VersionRange>`, and then a for loop to find a key that works for `javaV`, then checking `gradleV` against the value.

If we can't, we could make our very own minimal version of `VersionRange` to do the same thing.

Reply:
@christopherfujino suggested that I land this as is and not include `pub_semver`